node/pkg/cni/token_watch_test.go

```diff
@@ -0,0 +1,90 @@
+package cni_test
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/projectcalico/calico/node/pkg/cni"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var _ = Describe("FV tests", func() {
+	const serviceAccountName = "cni-token-watch-test"
+	const namespace = metav1.NamespaceSystem
+
+	It("should create a token successfully", func() {
+		clientset := createKubernetesClient()
+		setupServiceAccount(clientset, serviceAccountName, namespace)
+		defer cleanupServiceAccount(clientset, serviceAccountName, namespace)
+
+		tr := cni.NewTokenRefresher(clientset, namespace, serviceAccountName)
+		tu, err := tr.UpdateToken()
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(tu.Token).NotTo(BeEmpty())
+	})
+
+	It("should create a token successfully and deliver it through the channel", func() {
+		clientset := createKubernetesClient()
+		setupServiceAccount(clientset, serviceAccountName, namespace)
+		defer cleanupServiceAccount(clientset, serviceAccountName, namespace)
+
+		tr := cni.NewTokenRefresher(clientset, namespace, serviceAccountName)
+		tokenChan := tr.TokenChan()
+		go tr.Run()
+		tu := <-tokenChan
```

Comment:
This has the potential to hang for a long time, if the code is bugged.

I'd recommend switching the assertion like this:

```go

// Holder for the received token.
var receivedToken TokenUpdate

// Expect a token within a couple of seconds.
Eventually(bagelChan, 5*time.Second).Should(Receive(&receivedToken))

// Assert the received token is correct.
Expect(receivedToken.Token).NotTo(BeEmpty())
```

This way we should get a fast test failure if the token update takes longer than 5 seconds to be sent on the channel. 

Reference: https://onsi.github.io/gomega/#receive

Reply:
Thanks for the tip. Will be adapted accordingly.